Add 9P and remove SMB file sharing for Windows

redbeam opened the pull request:
> [!NOTE]
> Together with this PR, we also need to merge
> https://github.com/crc-org/snc/pull/1193 to add the 9pfs client to the bundle.

 - [x] https://github.com/crc-org/snc/pull/1193

----

This adds 9P file sharing support as discussed in #4858.

We are also removing SMB file sharing support for Windows in favor of 9P. Some of the work has already been done by @anjannath, and it is included here (#4620).

Fixes #4768

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Replaced SMB file sharing with 9P filesystem sharing on Windows.
  * Home directory now mounts automatically into the VM at startup.
  * Added redundant transport mechanism for improved reliability and compatibility.

* **Chores**
  * Updated dependencies to support 9P filesystem protocol.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cfergeau committed:
daemon: Use net.JoinHostPort

anjannath committed:
remove SMB file sharing

After the addition of 9p file sharing for Windows, we are removing
SMB file sharing support (#4768).

Part of this effort is to remove at least some of the warnings
reported by virustotal for the MSI:
https://www.virustotal.com/gui/file/31b402dcc1da24265074a21a26018d6cde8eef0b63c77a18f89eb079b6556790

redbeam committed:
add 9p file sharing

This commit adds 9P file sharing support to replace the existing
SMB file sharing solution (#3205, #3220, #4858).

cfergeau reviewed (commented):
Only took a quick look, but it’s unexpected to see the 9p code being added in the first commits before being disabled, and then to have the smb code removal.
Regarding the 9p code, some of the comments of coderabbit.ai (including in the vendor’ed code) deserve a closer look.

redbeam commented:
@cfergeau
> Only took a quick look, but it’s unexpected to see the 9p code being added in the first commits before being disabled, and then to have the smb code removal.

I based this branch on the 9p branch to avoid conflicts. This PR is expected to be merged as soon as 9p support is merged, so the 9p code will already be there.

> Regarding the 9p code, some of the comments of coderabbit.ai (including in the vendor’ed code) deserve a closer look.

I'll look at comments on my parts of the code, but what can be done about the vendored code?

cfergeau commented:
> I'll look at comments on my parts of the code, but what can be done about the vendored code?

If they are actual issues, in particular issues with security impacts, if it’s not too hard to fix them, I’d file PRs upstream, this way we’ll get the fixes in our codebase when they are merged.

redbeam commented on packaging/windows/product.wxs.template:
```diff
@@ -72,6 +66,11 @@
                             <RegistryValue Type="string" Name="ElementName" Value="gvisor-tap-vsock" KeyPath="yes"/>
                         </RegistryKey>
                     </Component>
+                    <Component Id="Hvsock9pRegistryEntry" Guid="*">
+                        <RegistryKey Root="HKLM" Key="Software\Microsoft\Windows NT\CurrentVersion\Virtualization\GuestCommunicationServices\00009000-FACB-11E6-BD58-64006A7986D3">
+                            <RegistryValue Type="string" Name="ElementName" Value="fs9p-hvsock" KeyPath="yes"/>
+                        </RegistryKey>
+                    </Component>
```
@anjannath do you think this is necessary? especially the "VsockRegistryEntry" one

anjannath replied:
@redbeam as per the docs, its only needed if the installer has two arches, for us its one arch, so it is fine, 
https://docs.firegiant.com/wix3/xsd/wix/component/

redbeam replied:
Alright, thanks! So this is just another CodeRabbit false positive :D

redbeam committed:
9p: hvsock support

This enables communication through hvsocks instead
of TCP sockets for 9P file sharing. TCP is still
utilized, as a backup option.

redbeam commented:
/hold

redbeam committed:
9p: add e2e feature test

anjannath commented:
@redbeam could you please squash the following 3 commits into one:

- [9p: add libraries and dependencies]
- [9p: initial version]
- [9p: add another shared directory for 9p when creating VM]

and also the following two commits into one:

- [remove enabling of file sharing and creation of smbshare from MSI]
- [remove SMB file sharing]


and add a descriptive commit log for the combined 9P commit, we need to also remove some of the logs from the smbshare removal commit (that creation of smbshare will be moved to `crc setup` which is not valid anymore)

I tested the PR and everything is working as expected, once this is updated, we are good to merge

redbeam commented:
@anjannath All done, please let me know if it looks okay :)

redbeam committed:
libhvee: make `ConvertToUnixPath` public and move it to a separate file

redbeam commented on test/e2e/testsuite/testsuite.go:
```diff
@@ -584,6 +586,19 @@ func InitializeScenario(s *godog.ScenarioContext) {
+	s.Step(`^basic file operations in mounted home directory should succeed$`,
+		func() error { return basicFileOperationsInDirectoryShouldSucceed(mountedHomeDir) })
+	s.Step(`^basic directory operations in mounted home directory should succeed$`,
+		func() error { return basicDirectoryOperationsInDirectoryShouldSucceed(mountedHomeDir) })
```
This function is now available for all platforms.

redbeam commented on test/e2e/testsuite/testsuite.go:
```diff
@@ -10,13 +10,15 @@ import (
 	"net/http"
 	"os"
 	"os/user"
+	"path"
```
This function is now available for all platforms.

redbeam commented on vendor/github.com/DeedleFake/p9/fs.go:
```diff
@@ -0,0 +1,729 @@
+	return QID{
+		Type: stat.FileMode.QIDType(),
+		Path: path,
+	}, nil
```
Fixed in https://github.com/DeedleFake/p9/pull/78

redbeam commented on vendor/github.com/DeedleFake/p9/addr_other.go:
```diff
@@ -0,0 +1,29 @@
+		panic(err)
+	}
+
+	display, ok := os.LookupEnv("DISPLAY")
+	if !ok {
+		display = ":0"
+	}
+
+	return filepath.Join(os.TempDir(), "ns."+u.Username+"."+display)
```
Fixed in https://github.com/DeedleFake/p9/pull/83/files

redbeam commented on vendor/github.com/DeedleFake/p9/dir.go:
```diff
@@ -0,0 +1,260 @@
+		}
+	}
+
+	name, ok := changes.Name()
+	if ok {
+		err := os.Rename(p, filepath.Join(base, filepath.FromSlash(name)))
+		if err != nil {
+			return err
+		}
+	}
```
Reported in https://github.com/DeedleFake/p9/issues/82

redbeam commented on pkg/fileserver/fs9p/server.go:
```diff
@@ -0,0 +1,98 @@
+	return nil
+}
+
+// WaitForError from a running server.
+func (s *Server) WaitForError() error {
+	err := <-s.ErrChan
+	// captures "accept tcp: endpoint is in invalid state" errors on exit
+	var opErr *net.OpError
+	if errors.As(err, &opErr) && strings.Contains(opErr.Error(), "endpoint is in invalid state") {
+		return nil
+	}
+	return err
+}
```
On other platforms, 9P is not going to be used :)

anjannath commented on test/e2e/testsuite/testsuite.go:
```diff
@@ -1363,3 +1378,115 @@ func getTimestamp(content string) error {
+	content := "test content"
+
+	// Create
+	_, err := util.SendCommandToVM(fmt.Sprintf("> %s", filename))
```
nit: in `zsh` it keeps wait for content on stdin to send to the file, maybe consider using `touch` instead?

anjannath replied:
nvm, on bash it works 👍🏽

anjannath replied:
aha so `touch` doesn't work:

```
touch: setting times of '/mnt/c/Users/rhqp/Desktop/anjan': Function not implemented
```

redbeam replied:
@anjannath yeah, this is the reason I had to avoid using `touch` ://

anjannath commented:
@redbeam `microshift` doesn't have a new release yet, and we need a new bundle of it with the `9pfs` client installed, otherwise crc start fails with the microshift preset

```
INFO Configuring shared directories
WARN Failed to connect to 9p server over hvsock: ssh command error:
command : 9pfs -V -p 36864 2 /mnt/c/Users/rhqp
err     : Process exited with status 127
WARN Falling back to 9p over TCP
ERRO ssh command error:
command : 9pfs 192.168.127.1 /mnt/c/Users/rhqp
err     : Process exited with status 127
```

we need to add a check to skip mounting shared directories when preset is microshift and the os is windows, wdyt?

redbeam committed:
9p: disable 9P filesharing for microshift until new bundles are released

anjannath commented:
@redbeam could you please also rebase it on main again

redbeam commented:
@anjannath I've addressed the microshift issue and rebased on the latest main.

redbeam commented on pkg/crc/machine/start.go:
```diff
@@ -248,15 +240,25 @@ func configureSharedDirs(vm *virtualMachine, sshRunner *crcssh.Runner) error {
+			if _, _, err := sshRunner.RunPrivileged("Changing owner of mount directory", "chown", "core:core", mount.Target); err != nil {
+				return err
+			}
+			if _, _, err := sshRunner.Run("9pfs -V -p", fmt.Sprintf("%d", constants.Plan9HvsockPort), "2", mount.Target); err != nil {
```
`2` is the vsock CID and means "host"

redbeam commented on pkg/crc/machine/start.go:
```diff
@@ -248,15 +240,25 @@ func configureSharedDirs(vm *virtualMachine, sshRunner *crcssh.Runner) error {
+			if _, _, err := sshRunner.RunPrivileged("Changing owner of mount directory", "chown", "core:core", mount.Target); err != nil {
+				return err
+			}
+			if _, _, err := sshRunner.Run("9pfs -V -p", fmt.Sprintf("%d", constants.Plan9HvsockPort), "2", mount.Target); err != nil {
+				logging.Warnf("Failed to connect to 9p server over hvsock: %v", err)
+				logging.Warnf("Falling back to 9p over TCP")
+				if _, _, err := sshRunner.Run("9pfs", constants.VSockGateway, mount.Target); err != nil {
+					return err
 				}
```
Not an issue.

redbeam commented:
/unhold

anjannath commented:
/lgtm
/approve
